feat(lambda,event-bridge): lambda event sources

thantos opened the pull request:
Closes #44 

Depends on #161 - This change made major changes to EventBus, stacking these changes to reduce churn.

* [x] Lambda
* [x] Step Functions - Already done
❌ Api Gateway
❌ App Sync
❌ Dynamo

 ## Lambda

Unlike Step Functions, Lambda does not emit events by default. Lambda has a concept of a `destination` or `async invocation`. Each lambda, version, and alias can only have a single onSuccess and onFailure destination. Destinations can be EventBuses, Sns topics, Sqs queues, or other lambda functions.

```ts
const func = new Function(stack, 'func', {
   onSuccess: bus, // IEventBus | IFunction | aws_lambda.IDestination
   onFailure: bus // IEventBus | IFunction | aws_lambda.IDestination
});

// fail when onSuccess is not set and not a bus?
func.onSuccess(bus).map().pipe();
func.onFailure(bus).map().pipe();
```

thantos commented on src/function.ts:
```diff
@@ -140,6 +208,56 @@ abstract class FunctionBase<P, O>
+      bus.bus,
+      id,
+      bus,
+      /**
+       * when(event => event.source === "lambda"
+       *    && event["detail-type"] === "Lambda Function Invocation Result - Failure"
+       *      && event.resources.includes(this.resource.functionArn))
+       */
```
yes :(

sam-goodwin commented on src/event-bridge/rule.ts:
```diff
@@ -138,8 +138,8 @@ export interface IRule<T extends Event> {
   pipe(func: IFunction<T, any>): void;
-  pipe(bus: IEventBus<T>): void;
-  pipe(props: EventBusTargetProps<T>): void;
+  pipe(bus: IEventBus<any>): void;
+  pipe(props: EventBusTargetProps<any>): void;
```
Why `any`? Wouldn't that break type safety?

sam-goodwin commented on src/function.ts:
```diff
@@ -52,8 +83,112 @@ export interface IFunction<P, O> {
+   *
+   * The rule returned will contain the logic:
+   *
+   * when(id, event => event.source === "lambda"
+   *    && event["detail-type"] === "Lambda Function Invocation Result - Success"
+   *      && event.resources.includes(this.resource.functionArn))
+   */
```
Surround with triple backticks block. Tsdoc produces markdown

sam-goodwin commented on src/function.ts:
```diff
@@ -140,6 +275,78 @@ abstract class FunctionBase<P, O>
+    });
+  }
+
+  onSuccess(
```
Does this inherit docs?

sam-goodwin replied:
No was just curious if it was happpening implicitly?

thantos replied:
Yeah, it inherits from the interface.

sam-goodwin commented on src/function.ts:
```diff
@@ -164,6 +374,9 @@ export interface FunctionProps
     clientName: ClientName | string
   ) => Omit<AWS.Lambda.ClientConfiguration, keyof AWS.Lambda.ClientApiVersions>;
+
+  onSuccess?: FunctionAsyncOnSuccessDestination<P, O>;
+  onFailure?: FunctionAsyncOnFailureDestination<P>;
```
Missing docs

sam-goodwin commented on src/function.ts:
```diff
@@ -140,6 +275,78 @@ abstract class FunctionBase<P, O>
     };
   }
+
+  protected static normalizeAsyncDestination<P, O>(
```
Why protected?

thantos replied:
This is an abstract base class, this method is used by the base and child class. And we don't need to expose this.

sam-goodwin commented on src/function.ts:
```diff
@@ -140,6 +275,78 @@ abstract class FunctionBase<P, O>
+      : destination;
+  }
+
+  configureAsyncInvoke(config: EventInvokeConfigOptions<P, O>): void {
```
Should this be public or private? I didn't see it in the interface so private?

thantos replied:
Its in the interface, made it public

![image](https://user-images.githubusercontent.com/289213/171630334-4f3fa650-cd7b-4c57-abc6-374f613c2be1.png)

thantos replied:
Added an es-lint rule to error on these cases to make your life easier.

thantos commented:
Note: prettier is broken due to the use of typescript 4.7

Waiting on: https://github.com/prettier/prettier/issues/12640

sam-goodwin commented on src/function.ts:
```diff
@@ -44,17 +57,154 @@ export type FunctionClosure<P, O> = (
+/**
+ * Returns the output type on the {@link IFunction}.
+ */
+ export type FunctionOutput<F extends IFunction<any, any>> = [F] extends [IFunction<any, infer O>] ? O : never;
```
nit: indentation

sam-goodwin replied:
why the `[F] extends `? What's the purpose of packing it a 1-tuple?

thantos replied:
`[F] extends []` changes the way that union types are interpreted.

https://www.typescriptlang.org/docs/handbook/2/conditional-types.html#distributive-conditional-types

With `[F]`

![image](https://user-images.githubusercontent.com/289213/172044197-5c567581-06ed-47b9-a030-4643409a967b.png)

Without

![image](https://user-images.githubusercontent.com/289213/172044243-98db4c3c-e616-4e64-a36a-1ac60535b6bb.png)

Which in the case its being used in... if the types are not blown out (`Function<A | B>`) then the `isFunction` does not narrow the types correctly. 

```ts
if(isFunction(destination)) {
    destination // is IFunction<AsyncResponseSuccess<P, O> | AsyncResponseFailure<P>, any>
} else {
     destination // is IDestination | IFunction<AsyncResponseSuccess<P, O> | AsyncResponseFailure<P>, any>
}
```

![image](https://user-images.githubusercontent.com/289213/172044280-d5638b53-48cc-40a4-af8e-1a0341166568.png)

But with... we get nice type narrowing.

```ts
if(isFunction(destination)) {
    destination // is IFunction<AsyncResponseSuccess<P, O>, any> | IFunction<AsyncResponseFailure<P>, any>
} else {
     destination // is IDestination
}
```

![image](https://user-images.githubusercontent.com/289213/172044366-e71f6af0-06a5-4d84-905c-8c256d71ac71.png)

sam-goodwin replied:
Awesome analysis. Thanks!

sam-goodwin commented on src/function.ts:
````diff
@@ -44,17 +57,154 @@ export type FunctionClosure<P, O> = (
+   *      && event.resources.includes(this.resource.functionArn))
+   * ```
+   */
+  onSuccess(
````
nit: looks like eslint missed this missing `public`

thantos replied:
Thats the interface :-)

sam-goodwin replied:
Ahhhh my bad. Thumbs down feels bad lol

sam-goodwin commented on src/function.ts:
```diff
@@ -44,17 +57,154 @@ export type FunctionClosure<P, O> = (
+  /**
+   * Event Source for the {@link Function} onSuccess async invocation destination.
+   *
+   * For Lambda, the onSuccess destination must first be configured via either
+   * the {@link Function} constructor or by using {@link IFunction.configureAsyncInvoke}
+   * and that destination must match the bus provided here.
```
I'm not sure what that means and how to take action on it.

sam-goodwin commented on src/event-bridge/event-bus.ts:
```diff
@@ -40,7 +40,12 @@ export const isEventBus = <E extends Event>(v: any): v is IEventBus<E> => {
+/**
+ * Returns the {@link Event} type on the {@link EventBus}.
+ */
+export type EventBusEvent<B extends IEventBus<any>> = [B] extends [IEventBus<infer E>] ? E : never;
```
Should this be called GetEventBusType? Or else it could be confused with the Event type. The verb helps with understanding

sam-goodwin commented on src/event-bridge/event-bus.ts:
```diff
@@ -115,16 +120,16 @@ export interface IEventBusFilterable<E extends Event> {
 }
 
-export interface IEventBus<E extends Event = Event>
+export interface IEventBus<in E extends Event = Event, O extends E = E>
```
Might want to document the type parameters. Not immediately obvious what E and O mean

sam-goodwin replied:
In standard libraries, I usually prefer to use words for type parameters instead of symbols.

thantos replied:
Thanks, that was a helpful exercise. 

```ts
/**
 * @typeParam Evnt - the union type of events that this EventBus can accept.
 *                   `Evnt` is the covariant version of `OutEnvt` in that
 *                   the bus will accept any of `Evnt` while the EventBus can
 *                   emit any of `OutEvnt`.
 * @typeParam OutEvnt - the union type of events that this EventBus will emit through rules.
 *                      `OutEvnt` is the contravariant version of `Envt` in that
 *                      the bus will emit any of `OutEvnt` while the EventBus can
 *                      can accept any of `Evnt`. This type parameter should be left
 *                      empty to be inferred.
 */
```

sam-goodwin commented on src/event-bridge/event-bus.ts:
```diff
@@ -339,7 +339,7 @@ abstract class EventBusBase<E extends Event> implements IEventBus<E> {
 
   public readonly eventBus = makeEventBusIntegration<
-    E,
+    EventBusPutEventInput<O>,
```
A bit verbose.

sam-goodwin commented on src/function.ts:
````diff
@@ -64,13 +116,153 @@ export interface IFunction<P, O>
+   *
+   * ```ts
+   * const bus = new EventBus(stack, 'bus');
+   * new Function(stack, 'func', { onSuccess: bus }, async () => {});
````
Cool!

sam-goodwin commented on src/function.ts:
```diff
@@ -64,13 +116,153 @@ export interface IFunction<P, O>
+   * const bus = new EventBus(stack, 'bus');
+   * const func = new Function(stack, 'func', async () => {});
+   * // if onSuccess or onFailure is already set, this will fail.
+   * func.configureAsyncInvoke({ onSuccess: bus });
```
Hmm configureAsyncInvoke isn't telling me exactly what it means. This is enabling async invoke and routing events to the bus? Maybe `enableAsyncInvoke`?

thantos replied:
This is the name from the CDK, which I agree, is also confusing..will take another pass.

sam-goodwin commented:
Should the title of the PR be more granular? Would someone reading the CHANGELOG know what this change is about based purely on the title?

thantos commented:
> Should the title of the PR be more granular? Would someone reading the CHANGELOG know what this change is about based purely on the title?

Ha... the change started with higher ambitions and then I realized that no one else but lambda and step function publish events!

Will update.

sam-goodwin commented on src/event-bridge/rule.ts:
```diff
@@ -21,11 +21,15 @@ import type { Event } from "./types";
+ * @typeParam - Evnt - The original event type from the {@link EventBus}.
+ * @typeParam - OutEvnt - The narrowed event type after the predicate is applied.
+ */
+export interface IRule<in Evnt extends Event, out OutEvnt extends Evnt> {
```
Why do we need two type parameters?

thantos replied:
`Evnt` is the event in. The input to the predicate on the rule.

`OutEvnt` is the narrowed type as the input to `pipe` or `map` or `when`.

Essentially splitting and showing the two sides of the predicate.

sam-goodwin replied:
Why do you need `Evnt`?

thantos replied:
Its being used on `integration: IntegrationWithEventBus<Evnt, Props>,`

But that doesn't seem right... will look.

sam-goodwin commented on src/event-bridge/event-bus.ts:
```diff
@@ -103,28 +108,42 @@ export interface IEventBusFilterable<E extends Event> {
-  when<O extends E>(
+    predicate: RulePredicateFunction<InEvnt, NewEvnt>
+  ): Rule<InEvnt, NewEvnt>;
+  when<InEvnt extends OutEvnt, NewEvnt extends InEvnt>(
```
These two parameters seem overly complex and redundant. Why did you add them?

thantos replied:
Need to invert the `in`Evnt` to `out NewEvnt` and `out OutEvnt` to `in InEvnt` for the when method.

sam-goodwin commented on src/event-bridge/event-bus.ts:
```diff
@@ -103,28 +108,42 @@ export interface IEventBusFilterable<E extends Event> {
-  extends IEventBusFilterable<E>,
+/**
+ * @typeParam Evnt - the union type of events that this EventBus can accept.
+ *                   `Evnt` is the covariant version of `OutEvnt` in that
```
`Evnt` is contra-variant, not co-variant. `out` is for covariance

sam-goodwin commented on src/event-bridge/event-bus.ts:
```diff
@@ -103,28 +108,42 @@ export interface IEventBusFilterable<E extends Event> {
+ *                      can accept any of `Evnt`. This type parameter should be left
+ *                      empty to be inferred. ex: `EventBus<Event<Detail1> | Event<Detail2>>`.
+ */
+export interface IEventBus<in Evnt extends Event = Event, OutEvnt extends Evnt = Evnt>
```
This doesn't seem right. I think we can remove the `OutEvnt`

sam-goodwin reviewed (approved):
We can modify the types in a subsequent change. Ship it

thantos commented:
> We can modify the types in a subsequent change. Ship it

Removed the unessecary `in Evnt` from `IRule` and `RuleBase` which cleaned up a bit.

Also fixed the covariance/contravariant naming issues and think I understand them better now :-)

https://dmitripavlutin.com/typescript-covariance-contravariance/

thantos commented:
> Error: Cannot find module '/opt/build/repo/scripts/compile-error-code-page'

I don't understand what changed
